DOC: Improve histogram2d() example.

wrwrwr opened the pull request:
A couple of small improvements:
- made the bins matrix non-square for a slightly more general example;
- unswapped variables in the histogram call (that only flipped the data coordinates);
- added a transposition of the resulting counts matrix (needed for all three plots);
- dropped the fixed histogram override (unnecessary and too symmetric to discern the orientation);
- turned some inessential matplotlib statements into keyword arguments;
- simplified expressions for xy-lims and centers;
- added links to matplotlib entities.

Here's a rendition of the example with the changes: [numpy-ref-p1139.pdf](https://github.com/numpy/numpy/files/510957/numpy-ref-p1139.pdf).

On a side note, how about using something like [extlinks](http://www.sphinx-doc.org/en/stable/ext/extlinks.html) (so the matplotlib links could be inlined).

See also #8115.

wrwrwr committed:
DOC: Improve histogram2d() example.

Closes #8115.

rgommers commented:
Thanks @wrwrwr 

> On a side note, how about using something like extlinks (so the matplotlib links could be inlined).

The right way to do this is to use ``intersphinx`` (http://www.sphinx-doc.org/en/stable/ext/intersphinx.html), we already use that to link to numpy and Python docs.  See ``doc/source/conf.py``. Could you check if that works for Matplotlib too (it should)?

wrwrwr committed:
DOC: Add matplotlib inventory for intersphinx.

This makes it possible to directly use matplotlib references such as
:func:`matplotlib.pyplot.imshow`.

wrwrwr commented:
@rgommers Thanks for the info, it does work.

Added matplotlib inventory to intersphinx configuration, while also changing `intersphinx_mapping` to the new format and adding an "s" to the python URI (to save a line of console output due to a redirect).
Updated the links, plus a minor rewording ("equidistant" &rarr; "square", "exact" &rarr; "actual"; [updated .pdf](https://github.com/numpy/numpy/files/517689/numpy-ref-1139.pdf)).


A usage tip. A convenient way to check what references are available:
```bash
wget matplotlib.org/objects.inv
python -m sphinx.ext.intersphinx objects.inv
```

rgommers commented:
> A usage tip. A convenient way to check what references are available:

thanks, that is indeed useful.

rgommers commented:
LGTM, merging. Thanks @wrwrwr
